jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java

```diff
@@ -0,0 +1,180 @@
+                    .build());
+        }
+        boundMenu.update(bot);
+    }
+
+    private final String loadingMessage;
+    private final MenuStateImpl initialState;
+    private final MenuStateMemoryImpl stateMemory;
+    private final MenuContext context;
+
+    public MenuImpl(@Nonnull String loadingMessage, @Nonnull Supplier<String> textSupplier, @Nullable ParseMode parseMode) {
+        this.loadingMessage = loadingMessage;
+        this.initialState = new MenuStateImpl(textSupplier, parseMode);
+        this.stateMemory = new MenuStateMemoryImpl(initialState);
+        this.context = new MenuContext();
+    }
+
+    @Nonnull
+    @Override
+    public String getLoadingMessage() {
+        return this.loadingMessage;
+    }
+
+    @Nonnull
+    @Override
+    public MenuStateImpl getInitialState() {
+        return this.initialState;
+    }
+
+    @Nonnull
+    @Override
+    public MenuStateMemoryImpl getStateMemory() {
+        return this.stateMemory;
+    }
+
+    @Nonnull
+    @Override
+    public MenuStateImpl getState() {
+        return Objects.requireNonNull(this.stateMemory.peekState(), "unexpected null state");
+    }
+
+    @Nonnull
+    @Override
+    public MenuContext getContext() {
+        return this.context;
+    }
+
+    @Nonnull
+    @Override
+    public MenuStateImpl createState(@Nonnull Supplier<String> textSupplier, @Nullable ParseMode parseMode) {
+        return new MenuStateImpl(textSupplier, parseMode);
+    }
+
+    @Nonnull
+    @Override
+    public MenuButtonImpl createButton(@Nonnull Consumer<InlineKeyboardButton.InlineKeyboardButtonBuilder> buttonCreator, @Nonnull OnClickHandler onClickHandler) {
+        return new MenuButtonImpl(buttonCreator, onClickHandler);
+    }
+
+    @Nonnull
+    @Override
+    public BoundMenuImpl send(@Nonnull TelegramBot bot, @Nonnull Chat chat, @Nullable Integer replyToMessageId) throws TelegramException {
+        AtomicReference<Object> messageOrError = new AtomicReference<>();
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        bot.perform(SendText.builder()
+                .chatId(ChatId.of(chat))
+                .replyToMessageID(replyToMessageId)
+                .text(getLoadingMessage())
+                .errorHandler(e -> {
+                    messageOrError.set(e);
+                    countDownLatch.countDown();
+                })
+                .callback(m -> {
+                    messageOrError.set(m);
+                    countDownLatch.countDown();
+                })
+                .build());
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException ex) {
+            throw new RuntimeException(ex);
+        }
+        Object moe = messageOrError.get();
+        if (moe instanceof TextMessage) {
+            TextMessage message = (TextMessage) moe;
+            BoundMenuImpl boundMenu = new BoundMenuImpl(this, message);
+            menusByMessageId.put(message.getMessageId(), boundMenu);
```

Comment:
Okay, so at first I thought this was a really smart idea until I remembered that message ids are not unique across different chats. IIRC They increment from a starting point and two messages can share the same message id given that they are from two different chats. Another solution is to add the chat id on top of that but... at this point I'd suggest, like before, having ids specified by the developer that uniquely identify the menu. To avoid any weird persistence issues, this would mean that the metadata thing you did here would have to be removed. If the IDs are specified by the developer, they can handle persistence across restarts.

IMO a developer shouldn't have to worry that restarting their bot will break their end users' experiences. It should be seamless and this system should support it.

Reply:
RIP I forgot about the lack of uniqueness.